framework/src/service/ServiceListeners.cpp

```diff
@@ -82,82 +93,170 @@ void ServiceListeners::Clear()
+}
+
+ListenerToken ServiceListeners::AddBundleListener(const std::shared_ptr<BundleContextPrivate>& context, const BundleListener& listener, void* data)
+{
+  auto token = MakeListenerToken();
+
+  auto l = bundleListenerMap.Lock(); US_UNUSED(l);
+  auto& listeners = bundleListenerMap.value[context];
+  listeners.push_back(std::make_tuple(listener, data, token.getId()));
+  return token;
 }
 
+/**
+ * Called by the deprecated RemoveBundleListener(name_of_callable)
+ */
 void ServiceListeners::RemoveBundleListener(const std::shared_ptr<BundleContextPrivate>& context, const BundleListener& listener, void* data)
 {
   auto l = bundleListenerMap.Lock(); US_UNUSED(l);
-  bundleListenerMap.value[context].remove_if(std::bind(BundleListenerCompare(), std::make_pair(listener, data), std::placeholders::_1));
+  auto& listeners = bundleListenerMap.value[context];
+  listeners.erase(std::remove_if(listeners.begin(),
+                                 listeners.end(),
+                                 std::bind(BundleListenerCompareName, listener, data, std::placeholders::_1)),
+                  listeners.end());
 }
 
-void ServiceListeners::AddFrameworkListener(const std::shared_ptr<BundleContextPrivate>& context, const FrameworkListener& listener, void* data)
+ListenerToken ServiceListeners::AddFrameworkListener(const std::shared_ptr<BundleContextPrivate>& context, const FrameworkListener& listener, void* data)
 {
+  auto token = MakeListenerToken();
+
   auto l = frameworkListenerMap.Lock(); US_UNUSED(l);
   auto& listeners = frameworkListenerMap.value[context];
-  if (std::find_if(listeners.begin(), listeners.end(), std::bind(FrameworkListenerCompare(), std::make_pair(listener, data), std::placeholders::_1)) == listeners.end())
-  {
-    listeners.push_back(std::make_pair(listener, data));
-  }
+  listeners.push_back(std::make_tuple(listener, data, token.getId()));
+  return token;
 }
 
+/**
+ * Called by the deprecated RemoveFrameworkListener(name_of_callable)
+ */
 void ServiceListeners::RemoveFrameworkListener(const std::shared_ptr<BundleContextPrivate>& context, const FrameworkListener& listener, void* data)
 {
   auto l = frameworkListenerMap.Lock(); US_UNUSED(l);
   auto& listeners = frameworkListenerMap.value[context];
   listeners.erase(std::remove_if(listeners.begin(),
                                  listeners.end(),
-                                 std::bind(FrameworkListenerCompare(), std::make_pair(listener, data), std::placeholders::_1)),
+                                 std::bind(FrameworkListenerCompareName, listener, data, std::placeholders::_1)),
                   listeners.end());
 }
 
+/**
+ * Called by RemoveListener(listenerToken)
+ * Iterates over the set of FrameworkListeners, BundleListeners and ServiceListeners to find the tokenId and
+ * remove the listener corresponding to the tokenId.
+ */
+void ServiceListeners::RemoveListener(const std::shared_ptr<BundleContextPrivate>& context, const ListenerToken& token)
+{
+  {
```

Comment:
Having to go through all the maps feels wasteful to me. At least, we should check the service listener map first because service listeners are more likely to be added / removed during the life time of a program.

If it is not too much work, it would be interesting to compare the performance implication when having lots of listeners registered (of different type) and we want to remove a single one. Comparison would be between 3.0 and your branch. I would say lets have 100 framework listeners, 1000 bundle listeners, and 10.000 service listeners. Then add and remove an additional one of each type in a loop (e.g. 1000 times) in a dedicated thread which is created 10 times (they should start looping only when all 10 threads have been started) and measure the average time for the addition and removal. Be sure to use a release build.

If there is a big performance hit, we might re-consider using a bit flag in the token id to store the listener type. In case you are too busy, I could help  out with this kind of testing.